### PR TITLE
readers/finder: add graphite.finders.remote.RemoteFinder

### PR DESCRIPTION
**Pull request by iksaif**
This patch moves *all* the specificities of the remote code
to graphite.finders.remote.RemoteFinder, which can now be disabled.

This also means the following changes:
* worker_pool/pool.py now support multiple thread pools
* storage.FetchQuery now has leaves_only which will allow optimizations in finders*
* Cleanup of the inflight requests/locks to avoid leaking in datalib
* Headers are always forwarded, including in find queries made for fetch queries.
* Finders can now declare themselves are 'local' (files are stored locally) or 'remote'.
* Added BaseReader and BaseFinder interfaces.

The only change in behavior is that now RemoteFinder has its own thread pool which isn't the same as the main thread pool. In practice it should not change anything.

### CODE REVIEW

**Comment by iksaif** [comment]
See https://github.com/graphite-project/graphite-web/issues/1943

Sorry for the super large PR, but it was super tricky to isolate the changes.

File: webapp/graphite/finders/remote.py [hunk 1]
```diff
@@ -11,48 +13,98 @@
...
+        log.debug("Got all find results in %fs" % (time.time() - start))
+
+    def prefetch(self, patterns, startTime, endTime, now, requestContext):
+        # Go through all of the remote nodes, and launch a fetch for each one.
+        # Each fetch will take place in its own thread, since it's naturally
+        # parallel work.
+        for store in self.remote_stores:
+            reader = RemoteReader(
+                store, {'intervals': []}, bulk_query=patterns)
         reader.fetch_list(startTime, endTime, now, requestContext)
```
Comment (DanCech) [inline webapp/graphite/finders/remote.py]: Shouldn't this be inside the loop?

File: webapp/graphite/render/datalib.py [hunk 2]
```diff
@@ -109,64 +109,24 @@ def getInfo(self):
...
-    result_queue = [
-      (node.path, node.fetch(startTime, endTime, now, requestContext))
-      for node in matching_nodes
-      if node.is_leaf
-    ]
-
-  log.info("render.datalib.fetchData :: starting to merge")
+  matching_nodes = STORE.find(
+    pathExpr, startTime, endTime,
+    local=requestContext['localOnly'],
```
Comment (DanCech) [inline webapp/graphite/render/datalib.py]: This seems like it's going to trigger find requests to all the remote hosts.  Avoiding those extra requests when using prefetch was part of the optimization we did to improve performance in 1.x

Comment (iksaif) [inline webapp/graphite/render/datalib.py]: right, I'll find a solution to avoid that

File: webapp/graphite/finders/remote.py (hunk 1 above)
Comment (iksaif) [inline webapp/graphite/finders/remote.py]: good catch

File: webapp/graphite/render/datalib.py (hunk 2 above)
Comment (iksaif) [inline webapp/graphite/render/datalib.py]: Ok, totally forgot that. Fixed

File: webapp/graphite/render/views.py [hunk 3]
```diff
@@ -48,6 +50,36 @@
...
+  each time evaluateTarget() needs to fetch a path. All the prefetched
+  data is stored in the requestContext, to be accessed later by datalib.
+  """
+  log.rendering("Prefetching remote data")
+  pathExpressions = extractPathExpressions(targets)
+  results = STORE.prefetch(requestContext, pathExpressions)
+
+  prefetched = collections.defaultdict(list)
+  for result in results:
+    fetched = wait_for_result(result)
```
Comment (DanCech) [inline webapp/graphite/render/views.py]: We want to wait for these results in the main loop, so that they don't block any local finds and fetches from being kicked off

Comment (iksaif) [inline webapp/graphite/render/views.py]: Right, I assumed this didn't really matter because we rarely mix local and remote together, but I'll make sure we do that as late as we can. New patch coming on monday :)

**Comment by iksaif** [comment]
Ok, how does it look like now ?

This is still some potential for cleanups but I'd like to get that in first. Some ideas for future cleanups:

* Use concurrent.futures for ThreadPool and futures (this would give us better abstraction for futures).
* Double check all the usages of `FetchInProgress`

**Comment by DanCech** [comment]
Looking good, I'll do some testing as soon as I can.  @replay how does this look to you?

File: webapp/graphite/finders/remote.py [hunk 4]
```diff
@@ -11,48 +13,124 @@
...
+        # parallel work.
+
+        nodes = []
+        patterns = []
+
+        for v in nodes_or_patterns:
+            if isinstance(v, basestring):
+                patterns.append(v)
+            else:
+                nodes.append(v)
```
Comment (replay) [inline webapp/graphite/finders/remote.py]: is this used anywhere?

**Comment by replay** [comment]
This looks really good, still going through it in detail

File: webapp/graphite/future.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+"""Lightweight Future implementation.
+
+Before doing that we should evalluate the use of concurrent.futures
```
Comment (replay) [inline webapp/graphite/future.py]: `evaluate`

File: webapp/graphite/future.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+first, which could also replace FetchInProgress.
+"""
+import abc
+
+from graphite.logger import log
+
+
+class Future(object) :
+    def __init__(self, *args, **kwargs) :
+        super(Future, self).__init__(*args, **kwargs)
```
Comment (replay) [inline webapp/graphite/future.py]: I'm not sure if i just don't get what's happening here, but doesn't that just always call `__init__()` on the `object` it's inheriting from? what does that do?

Comment (iksaif) [inline webapp/graphite/future.py]: I agree that calling super() on object isn't super useful, I'll remove that

File: webapp/graphite/future.py (hunk 5 above)
Comment (iksaif) [inline webapp/graphite/future.py]: done

File: webapp/graphite/finders/remote.py (hunk 4 above)
Comment (iksaif) [inline webapp/graphite/finders/remote.py]: Oh right, fixed.

**Comment by iksaif** [comment]
Just sent a bunch of fixes and added a few unit tests.

**Comment by iksaif** [comment]
So, what do you all think ? I'd merge that to master, and then merge all the potential regressions before the next release.

**Comment by deniszh** [comment]
Looks good, @iksaif !
Please go on! 👍

**Comment by deniszh** [comment]
I think we should release 1.1.0 after such massive change btw.

**Comment by DanCech** [comment]
I haven't actually had a chance to test this yet, @deniszh did you do any testing?

**Comment by deniszh** [comment]
No, not yet. Will try tomorrow. But proper test requires quite a bit cluster, it's hard to cover all corner cases with simple test setup...
